File: src/agent-handle.ts
/**
 * AgentHandle - Represents a running agent with an ACP connection
 */

import { spawn, type ChildProcess } from "node:child_process";
import { Writable, Readable } from "node:stream";
import * as acp from "@agentclientprotocol/sdk";
import type { AgentConfig, SpawnOptions, SessionOptions, ForkSessionOptions } from "./types.js";
import type { AgentCapabilities } from "@agentclientprotocol/sdk";
import { Session } from "./session.js";
import { ACPClientHandler } from "./client-handler.js";

/**
 * Handle to a running agent process with ACP connection
 */
export class AgentHandle {
  readonly capabilities: AgentCapabilities;

  /**
   * Map of session IDs to Session objects for tracking active sessions.
   * Used for smart fork detection to determine if a session is processing.
   */
  private readonly sessions: Map<string, Session> = new Map();

  private constructor(
    private readonly process: ChildProcess,
    private readonly connection: acp.ClientSideConnection,
    private readonly clientHandler: ACPClientHandler,
    capabilities: AgentCapabilities
  ) {
    this.capabilities = capabilities;
  }

  /**
   * Create and initialize an agent handle
   * @internal
   */
  static async create(
    config: AgentConfig,
    options: SpawnOptions
  ): Promise<AgentHandle> {
    // 1. Spawn subprocess
    const env = {
      ...process.env,
      ...config.env,
      ...options.env,
    };

    const agentProcess = spawn(config.command, config.args, {
      stdio: ["pipe", "pipe", "inherit"],
      env,
    });

    // Handle process errors
    agentProcess.on("error", (err) => {
      console.error(`Agent process error: ${err.message}`);
    });

    // 2. Set up NDJSON streams
    if (!agentProcess.stdin || !agentProcess.stdout) {
      agentProcess.kill();
      throw new Error("Failed to get agent process stdio streams");
    }

    const input = Writable.toWeb(agentProcess.stdin);
    const output = Readable.toWeb(agentProcess.stdout) as ReadableStream<Uint8Array>;
    const stream = acp.ndJsonStream(input, output);

    // 3. Create client handler and connection
    const clientHandler = new ACPClientHandler(
      {
        onPermissionRequest: options.onPermissionRequest,
        onFileRead: options.onFileRead,
        onFileWrite: options.onFileWrite,
        onTerminalCreate: options.onTerminalCreate,
        onTerminalOutput: options.onTerminalOutput,
        onTerminalKill: options.onTerminalKill,
        onTerminalRelease: options.onTerminalRelease,
        onTerminalWaitForExit: options.onTerminalWaitForExit,
      },
      options.permissionMode ?? "auto-approve"
    );

    const connection = new acp.ClientSideConnection(
      () => clientHandler,
      stream
    );

    // 4. Initialize connection
    try {
      const initResult = await connection.initialize({
        protocolVersion: acp.PROTOCOL_VERSION,
        clientCapabilities: {
          fs: {
            readTextFile: true,
            writeTextFile: true,
          },
          terminal: !!(
            options.onTerminalCreate &&
            options.onTerminalOutput &&
            options.onTerminalKill &&
            options.onTerminalRelease &&
            options.onTerminalWaitForExit
          ),
        },
      });

      return new AgentHandle(
        agentProcess,
        connection,
        clientHandler,
        initResult.agentCapabilities ?? {}
      );
    } catch (error) {
      agentProcess.kill();
      throw error;
    }
  }

  /**
   * Create a new session with the agent
   */
  async createSession(
    cwd: string,
    options: SessionOptions = {}
  ): Promise<Session> {
    const result = await this.connection.newSession({
      cwd,
      mcpServers: options.mcpServers ?? [],
    });

    // Set mode if specified
    if (options.mode && this.connection.setSessionMode) {
      await this.connection.setSessionMode({
        sessionId: result.sessionId,
        modeId: options.mode,
      });
    }

    const session = new Session(
      result.sessionId,
      this.connection,
      this.clientHandler,
      cwd,
      result.modes?.availableModes?.map((m: { id: string }) => m.id) ?? [],
      result.models?.availableModels?.map((m: { modelId: string }) => m.modelId) ?? []
    );

    // Track session for smart fork detection
    this.sessions.set(result.sessionId, session);

    return session;
  }

  /**
   * Load an existing session by ID
   */
  async loadSession(
    sessionId: string,
    cwd: string,
    mcpServers: Array<{ name: string; uri: string }> = []
  ): Promise<Session> {
    if (!this.capabilities.loadSession) {
      throw new Error("Agent does not support loading sessions");
    }

    const result = await this.connection.loadSession({
      sessionId,
      cwd,
      mcpServers,
    });

    const session = new Session(
      sessionId,
      this.connection,
      this.clientHandler,
      cwd,
      result.modes?.availableModes?.map((m: { id: string }) => m.id) ?? [],
      result.models?.availableModels?.map((m: { modelId: string }) => m.modelId) ?? []
    );

    // Track session for smart fork detection
    this.sessions.set(sessionId, session);

    return session;
  }

  /**
   * Fork an existing session to create a new independent session
   *
   * The forked session inherits the conversation history from the original,
   * allowing operations without affecting the original session's state.
   *
   * This method uses smart detection to determine the best forking approach:
   * - If the source session is actively processing or not persisted, uses forkWithFlush
   * - If the source session is idle and persisted, uses direct fork
   * - Use `forceFlush: true` to always use the flush approach
   *
   * @param sessionId - The ID of the session to fork
   * @param cwd - The current working directory for the forked session
   * @param options - Optional fork configuration
   * @experimental This relies on the unstable session/fork ACP capability
   */
  async forkSession(
    sessionId: string,
    cwd: string,
    options: ForkSessionOptions = {}
  ): Promise<Session> {
    if (!this.capabilities.sessionCapabilities?.fork) {
      throw new Error("Agent does not support forking sessions");
    }

    const sourceSession = this.sessions.get(sessionId);

    // Determine if flush is needed:
    // 1. forceFlush option is set
    // 2. Source session is currently processing
    // 3. Source session is not tracked (might be from previous process, needs flush to ensure persistence)
    const needsFlush = options.forceFlush ||
      (sourceSession && sourceSession.isProcessing) ||
      !sourceSession;

    if (needsFlush && sourceSession) {
      // Use forkWithFlush for active or processing sessions
      const forkedSession = await sourceSession.forkWithFlush({
        idleTimeout: options.idleTimeout,
        persistTimeout: options.persistTimeout,
      });

      // Track the forked session
      this.sessions.set(forkedSession.id, forkedSession);

      return forkedSession;
    }

    // Direct fork for persisted idle sessions (or when source session is unknown)
    const result = await this.connection.unstable_forkSession({
      sessionId,
      cwd,
    });

    const forkedSession = new Session(
      result.sessionId,
      this.connection,
      this.clientHandler,
      cwd,
      result.modes?.availableModes?.map((m: { id: string }) => m.id) ?? [],
      result.models?.availableModels?.map((m: { modelId: string }) => m.modelId) ?? []
    );

    // Track the forked session
    this.sessions.set(result.sessionId, forkedSession);

    return forkedSession;
  }

  /**
   * Close the agent connection and terminate the process
   */
  async close(): Promise<void> {
    this.process.kill();
    // Wait for the connection to close
    await this.connection.closed;
  }

  /**
   * Get the underlying connection (for advanced use)
   */
  getConnection(): acp.ClientSideConnection {
    return this.connection;
  }

  /**
   * Check if the agent process is still running
   */
  isRunning(): boolean {
    return !this.process.killed && this.process.exitCode === null;
  }
}

File: python/src/acp_factory/agent_handle.py
"""
AgentHandle - Represents a running agent with an ACP connection.
"""

from __future__ import annotations

import os
from asyncio.subprocess import Process
from contextlib import AsyncExitStack
from typing import TYPE_CHECKING, Any

import acp
from acp.client import ClientSideConnection
from acp.schema import ClientCapabilities, FileSystemCapability

from acp_factory.client_handler import ACPClientHandler
from acp_factory.types import (
    AgentConfig,
    ForkSessionOptions,
    SessionOptions,
    SpawnOptions,
)

if TYPE_CHECKING:
    from acp_factory.session import Session


class AgentHandle:
    """
    Handle to a running agent process with ACP connection.

    This class manages the lifecycle of an agent subprocess and provides
    methods to create and manage sessions with the agent.
    """

    def __init__(
        self,
        process: Process,
        connection: ClientSideConnection,
        client_handler: ACPClientHandler,
        capabilities: dict[str, Any],
        exit_stack: AsyncExitStack,
    ) -> None:
        self._process = process
        self._connection = connection
        self._client_handler = client_handler
        self.capabilities = capabilities
        self._sessions: dict[str, Session] = {}
        self._exit_stack = exit_stack

    @classmethod
    async def create(
        cls,
        config: AgentConfig,
        options: SpawnOptions,
    ) -> AgentHandle:
        """Create and initialize an agent handle."""
        from acp_factory.session import Session  # noqa: F401

        # Merge environment variables
        env = {**os.environ, **config.env, **options.env}

        # Create client handler
        client_handler = ACPClientHandler(
            handlers=options,
            permission_mode=options.permission_mode,
        )

        # Determine terminal capability based on handlers
        has_terminal = all([
            options.on_terminal_create,
            options.on_terminal_output,
            options.on_terminal_kill,
            options.on_terminal_release,
            options.on_terminal_wait_for_exit,
        ])

        # Use ACP's spawn_agent_process for proper stream handling
        exit_stack = AsyncExitStack()
        try:
            connection, process = await exit_stack.enter_async_context(
                acp.spawn_agent_process(
                    client_handler,
                    config.command,
                    *config.args,
                    env=env,
                    use_unstable_protocol=True,
                )
            )

            # Initialize connection
            init_result = await connection.initialize(
                protocol_version=acp.PROTOCOL_VERSION,
                client_capabilities=ClientCapabilities(
                    fs=FileSystemCapability(
                        read_text_file=True,
                        write_text_file=True,
                    ),
                    terminal=has_terminal,
                ),
            )

            capabilities = init_result.agent_capabilities or {}
            # Convert to dict if it's a pydantic model
            if hasattr(capabilities, "model_dump"):
                capabilities = capabilities.model_dump()
            elif hasattr(capabilities, "dict"):
                capabilities = capabilities.dict()
            elif not isinstance(capabilities, dict):
                capabilities = dict(capabilities) if capabilities else {}

            return cls(process, connection, client_handler, capabilities, exit_stack)

        except Exception as e:
            await exit_stack.aclose()
            raise e

    async def create_session(
        self,
        cwd: str,
        options: SessionOptions | None = None,
    ) -> Session:
        """Create a new session with the agent."""
        from acp_factory.session import Session

        options = options or SessionOptions()

        result = await self._connection.new_session(
            cwd=cwd,
            mcp_servers=options.mcp_servers or [],
        )

        # Set mode if specified
        if options.mode:
            await self._connection.set_session_mode(
                session_id=result.session_id,
                mode_id=options.mode,
            )

        # Extract modes and models from result
        modes: list[str] = []
        models: list[str] = []

        if result.modes and result.modes.available_modes:
            modes = [m.id for m in result.modes.available_modes]

        if result.models and result.models.available_models:
            models = [m.model_id for m in result.models.available_models]

        session = Session(
            session_id=result.session_id,
            connection=self._connection,
            client_handler=self._client_handler,
            cwd=cwd,
            modes=modes,
            models=models,
        )

        # Track session for smart fork detection
        self._sessions[result.session_id] = session

        return session

    async def load_session(
        self,
        session_id: str,
        cwd: str,
        mcp_servers: list[dict[str, str]] | None = None,
    ) -> Session:
        """Load an existing session by ID."""
        from acp_factory.session import Session

        mcp_servers = mcp_servers or []

        # Check capability
        load_session_cap = self.capabilities.get("loadSession", False)
        if not load_session_cap:
            raise RuntimeError("Agent does not support loading sessions")

        result = await self._connection.load_session(
            session_id=session_id,
            cwd=cwd,
            mcp_servers=mcp_servers,
        )

        # Extract modes and models from result
        modes: list[str] = []
        models: list[str] = []

        if result.modes and result.modes.available_modes:
            modes = [m.id for m in result.modes.available_modes]

        if result.models and result.models.available_models:
            models = [m.model_id for m in result.models.available_models]

        session = Session(
            session_id=session_id,
            connection=self._connection,
            client_handler=self._client_handler,
            cwd=cwd,
            modes=modes,
            models=models,
        )

        # Track session for smart fork detection
        self._sessions[session_id] = session

        return session

    async def fork_session(
        self,
        session_id: str,
        cwd: str,
        options: ForkSessionOptions | None = None,
    ) -> Session:
        """
        Fork an existing session to create a new independent session.

        The forked session inherits the conversation history from the original,
        allowing operations without affecting the original session's state.
        """
        from acp_factory.session import Session

        options = options or ForkSessionOptions()

        # Check capability
        session_caps = self.capabilities.get("sessionCapabilities", {})
        if not session_caps.get("fork"):
            raise RuntimeError("Agent does not support forking sessions")

        source_session = self._sessions.get(session_id)

        # Determine if flush is needed
        needs_flush = (
            options.force_flush
            or (source_session is not None and source_session.is_processing)
            or source_session is None
        )

        if needs_flush and source_session is not None:
            # Use fork_with_flush for active or processing sessions
            forked_session = await source_session.fork_with_flush(
                idle_timeout=options.idle_timeout,
                persist_timeout=options.persist_timeout,
            )
            self._sessions[forked_session.id] = forked_session
            return forked_session

        # Direct fork for persisted idle sessions
        result = await self._connection.fork_session(
            session_id=session_id,
            cwd=cwd,
        )

        # Extract modes and models from result
        modes: list[str] = []
        models: list[str] = []

        if result.modes and result.modes.available_modes:
            modes = [m.id for m in result.modes.available_modes]

        if result.models and result.models.available_models:
            models = [m.model_id for m in result.models.available_models]

        forked_session = Session(
            session_id=result.session_id,
            connection=self._connection,
            client_handler=self._client_handler,
            cwd=cwd,
            modes=modes,
            models=models,
        )

        # Track the forked session
        self._sessions[result.session_id] = forked_session

        return forked_session

    async def close(self) -> None:
        """Close the agent connection and terminate the process."""
        await self._exit_stack.aclose()

    def get_connection(self) -> ClientSideConnection:
        """Get the underlying connection (for advanced use)."""
        return self._connection

    def is_running(self) -> bool:
        """Check if the agent process is still running."""
        return self._process.returncode is None
